scripts/generate-types.js

```diff
@@ -0,0 +1,72 @@
+const path = require('path');
+const { EOL } = require('os');
+const fs = require('fs').promises;
+const { getComponentNames, constants, compStateLogger } = require('./helpers');
+
+function capitalizeFirstLetter(string) {
+	return string.charAt(0).toUpperCase() + string.slice(1);
+}
+
+const camelize = s => s.replace(/-./g, x => x[1].toUpperCase());
+
+const generatePropertyLine = (name, type, description) => {
+	let result = '';
+	if (description) {
+		result += `\t/** ${description} */${EOL}`;
+	}
+	if (name.includes('-')) {
+		// if name has dashes we put it in quotes
+		name = `"${name}"`;
+	}
+	// Replace 'Array' with '[]'.
+	// Better to use Array<T> but we don't have information about structure of array items now.
+	if (type.toLowerCase() === 'array') {
+		type = '[]';
+	}
+	result += `\t${name}?: ${type.toLowerCase()};${EOL}${EOL}`;
+	return result;
+};
+
+/**
+ *  Read description from docs/properties and add it to the src file
+ *  Single time script to update the src file with all available description in README files
+ */
+(async function () {
+	// read the data from the src properties
+	// get the component names
+	const componentNames = getComponentNames();
+	for (const compName of componentNames) {
+		// get the component src properties
+		compStateLogger(compName, 'Reading src/properties.json ...');
+		const componentDir = path.join(process.cwd(), `${constants.componentsPath}/${compName}`);
+		const srcProps = await fs.readFile(`${componentDir}/docs/src/properties.json`, 'utf8');
+		const properties = JSON.parse(srcProps).filter(
+			// filter empty properties and internal properties
+			property => property.Property && !property.Description.includes('INTERNAL')
+		);
+
+		const typesFilePath = `${componentDir}/types/${compName}.d.ts`;
+		await fs.mkdir(`${componentDir}/types/`, { recursive: true });
+		const className = camelize(capitalizeFirstLetter(compName));
+		let typesFileContent = '';
+
+		// Generate interface for html attributes
+		typesFileContent += `export interface TS${className}HTMLAttributes {${EOL}`;
+		properties.forEach(async property => {
+			typesFileContent += generatePropertyLine(property.Attribute, property.Type, property.Description);
+		});
+
+		typesFileContent += `}${EOL}${EOL}`;
+
+		// Generate interface for js properties
+		typesFileContent += `export interface TS${className} {${EOL}`;
+		properties.forEach(async property => {
+			typesFileContent += generatePropertyLine(property.Property, property.Type, property.Description);
+		});
+
+		typesFileContent += `}${EOL}`;
+
+		await fs.writeFile(typesFilePath, typesFileContent);
+		compStateLogger(compName, 'Added type definitions.');
```

Comment:
nit, I think we can make some smaller functions like `getExposedProperties`, `getHtmlAttributeInterfaces`, `getJsPropertiesInterfaces`, `getTypesFileContent`, ... to make it even cleaner.

Reply:
Yeah, there's a tradeoff, I think I'm still high on uncle Bob clean code talk :D

Reply:
but in general I kinda like to get a general understanding of what is happening without the detail in the first look and if I cared I can find what is happening underneath. Although it would make it easier to test and maintain it in long term. But I understand how during the implementation it can feel too much, since you have to jump all over the place 👍

Reply:
In this particular case i think it is a grey area, because the function is big, so it is ok to have a feeling to split it in smaller ones. But this is also a script, not an actual function that is used in the project, so it is kinda standalone, so I think there is no clear answer, what is the right way to do in such cases :)

Reply:
Yeah, agree! :)

Reply:
Agree, especially if these functions can be expected to be used only here 👍